File: internal/cli/format.go
package cli

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"io"
	"sort"
	"strings"
)

func validateFormat(format string) error {
	switch format {
	case "table", "json", "csv":
		return nil
	default:
		return fmt.Errorf("unsupported format %q", format)
	}
}

func printResponse(writer io.Writer, format string, response engineResponse) error {
	switch format {
	case "json":
		return printJSON(writer, response)
	case "csv":
		return printCSV(writer, response)
	default:
		printTable(writer, response)
		return nil
	}
}

func printJSON(writer io.Writer, response engineResponse) error {
	rows := make([]map[string]interface{}, 0, len(response.Rows))
	for _, row := range response.Rows {
		item := make(map[string]interface{}, len(response.Columns))
		for i, column := range response.Columns {
			item[column] = row[i]
		}
		rows = append(rows, item)
	}

	encoder := json.NewEncoder(writer)
	return encoder.Encode(rows)
}

func printCSV(writer io.Writer, response engineResponse) error {
	csvWriter := csv.NewWriter(writer)
	if err := csvWriter.Write(response.Columns); err != nil {
		return err
	}
	for _, row := range response.Rows {
		record := make([]string, 0, len(row))
		for _, value := range row {
			record = append(record, fmt.Sprint(value))
		}
		if err := csvWriter.Write(record); err != nil {
			return err
		}
	}
	csvWriter.Flush()
	return csvWriter.Error()
}

func printTable(writer io.Writer, response engineResponse) {
	if len(response.Columns) == 0 {
		return
	}

	widths := make([]int, len(response.Columns))
	for i, column := range response.Columns {
		widths[i] = len(column)
	}
	for _, row := range response.Rows {
		for i, value := range row {
			cell := fmt.Sprint(value)
			if len(cell) > widths[i] {
				widths[i] = len(cell)
			}
		}
	}

	for i, column := range response.Columns {
		fmt.Fprintf(writer, "%-*s", widths[i], column)
		if i < len(response.Columns)-1 {
			fmt.Fprint(writer, "  ")
		}
	}
	fmt.Fprintln(writer)

	for i, width := range widths {
		fmt.Fprint(writer, strings.Repeat("-", width))
		if i < len(widths)-1 {
			fmt.Fprint(writer, "  ")
		}
	}
	fmt.Fprintln(writer)

	for _, row := range response.Rows {
		for i, value := range row {
			fmt.Fprintf(writer, "%-*v", widths[i], value)
			if i < len(row)-1 {
				fmt.Fprint(writer, "  ")
			}
		}
		fmt.Fprintln(writer)
	}
}

func supportedLanguages() []string {
	langs := []string{"go"}
	sort.Strings(langs)
	return langs
}

File: crates/ql-cli/src/format.rs
use std::io::Write;

use ql_adapters::supported_languages as adapter_supported_languages;
use ql_core::protocol::QueryResult;

const MAX_CELL_WIDTH: usize = 60;

pub fn validate_format(format: &str) -> Result<(), String> {
    match format {
        "table" | "json" | "csv" => Ok(()),
        _ => Err(format!("unsupported format {format:?}")),
    }
}

pub fn supported_languages() -> Vec<String> {
    adapter_supported_languages()
}

pub fn format_response(
    writer: &mut impl Write,
    format: &str,
    result: &QueryResult,
) -> Result<(), String> {
    match format {
        "json" => format_json(writer, result),
        "csv" => format_csv(writer, result),
        _ => format_table(writer, result),
    }
}

fn format_json(writer: &mut impl Write, result: &QueryResult) -> Result<(), String> {
    let rows: Vec<serde_json::Value> = result
        .rows
        .iter()
        .map(|row| {
            let mut item = serde_json::Map::new();
            for (i, column) in result.columns.iter().enumerate() {
                item.insert(column.clone(), row[i].clone());
            }
            serde_json::Value::Object(item)
        })
        .collect();

    let output = serde_json::to_string(&rows).map_err(|e| e.to_string())?;
    writeln!(writer, "{output}").map_err(|e| e.to_string())
}

fn format_csv(writer: &mut impl Write, result: &QueryResult) -> Result<(), String> {
    let mut csv = String::new();

    csv.push_str(&result.columns.join(","));
    csv.push('\n');

    for row in &result.rows {
        let record: Vec<String> = row
            .iter()
            .map(|value| match value {
                serde_json::Value::String(s) => {
                    if s.contains(',') || s.contains('"') || s.contains('\n') {
                        format!("\"{}\"", s.replace('"', "\"\""))
                    } else {
                        s.clone()
                    }
                }
                other => other.to_string(),
            })
            .collect();
        csv.push_str(&record.join(","));
        csv.push('\n');
    }

    write!(writer, "{csv}").map_err(|e| e.to_string())
}

fn display_cell(value: &serde_json::Value) -> String {
    let text = match value {
        serde_json::Value::String(s) => s.clone(),
        serde_json::Value::Null => String::new(),
        other => other.to_string(),
    };

    let mut chars = text.chars();
    let mut shortened: String = chars.by_ref().take(MAX_CELL_WIDTH).collect();
    if chars.next().is_some() {
        shortened.truncate(MAX_CELL_WIDTH.saturating_sub(3));
        shortened.push_str("...");
    }

    shortened
}

fn numeric_column(result: &QueryResult, index: usize) -> bool {
    let mut has_number = false;

    for row in &result.rows {
        match row.get(index) {
            Some(serde_json::Value::Null) => {}
            Some(serde_json::Value::Number(_)) => has_number = true,
            Some(_) | None => return false,
        }
    }

    has_number
}

fn format_table(writer: &mut impl Write, result: &QueryResult) -> Result<(), String> {
    if result.columns.is_empty() {
        return Ok(());
    }

    let mut widths: Vec<usize> = result.columns.iter().map(|column| column.len()).collect();
    let numeric = (0..result.columns.len())
        .map(|index| numeric_column(result, index))
        .collect::<Vec<_>>();

    for row in &result.rows {
        for (index, value) in row.iter().enumerate() {
            let cell = display_cell(value);
            widths[index] = widths[index].max(cell.len());
        }
    }

    for (i, column) in result.columns.iter().enumerate() {
        write!(writer, "{:<width$}", column, width = widths[i]).map_err(|e| e.to_string())?;
        if i < result.columns.len() - 1 {
            write!(writer, "  ").map_err(|e| e.to_string())?;
        }
    }
    writeln!(writer).map_err(|e| e.to_string())?;

    for (i, width) in widths.iter().enumerate() {
        write!(writer, "{}", "-".repeat(*width)).map_err(|e| e.to_string())?;
        if i < widths.len() - 1 {
            write!(writer, "  ").map_err(|e| e.to_string())?;
        }
    }
    writeln!(writer).map_err(|e| e.to_string())?;

    for row in &result.rows {
        for (i, value) in row.iter().enumerate() {
            let cell = display_cell(value);
            if numeric[i] {
                write!(writer, "{:>width$}", cell, width = widths[i])
            } else {
                write!(writer, "{:<width$}", cell, width = widths[i])
            }
            .map_err(|e| e.to_string())?;
            if i < row.len() - 1 {
                write!(writer, "  ").map_err(|e| e.to_string())?;
            }
        }
        writeln!(writer).map_err(|e| e.to_string())?;
    }

    Ok(())
}

pub fn clear_screen() {
    print!("\x1b[H\x1b[2J");
}

#[cfg(test)]
mod tests {
    use super::*;
    use ql_core::protocol::QueryResult;
    use serde_json::Value;

    #[test]
    fn validates_known_formats() {
        for fmt in &["table", "json", "csv"] {
            assert!(validate_format(fmt).is_ok());
        }
        assert!(validate_format("xml").is_err());
    }

    #[test]
    fn formats_table_output() {
        let result = QueryResult {
            columns: vec!["name".to_string(), "line".to_string()],
            rows: vec![
                vec![Value::String("main".to_string()), Value::from(4)],
                vec![Value::String("add".to_string()), Value::from(12)],
            ],
        };

        let mut output = Vec::new();
        format_response(&mut output, "table", &result).expect("format should succeed");

        let expected = "name  line\n----  ----\nmain     4\nadd     12\n";
        assert_eq!(String::from_utf8(output).unwrap(), expected);
    }

    #[test]
    fn truncates_long_cells() {
        let result = QueryResult {
            columns: vec!["description".to_string()],
            rows: vec![vec![Value::String("a".repeat(70))]],
        };

        let mut output = Vec::new();
        format_response(&mut output, "table", &result).expect("format should succeed");

        let output = String::from_utf8(output).unwrap();
        assert!(output.contains(&format!("{}...", "a".repeat(57))));
    }

    #[test]
    fn formats_json_output() {
        let result = QueryResult {
            columns: vec!["name".to_string(), "line".to_string()],
            rows: vec![vec![Value::String("main".to_string()), Value::from(4)]],
        };

        let mut output = Vec::new();
        format_response(&mut output, "json", &result).expect("format should succeed");

        let expected = "[{\"line\":4,\"name\":\"main\"}]\n";
        assert_eq!(String::from_utf8(output).unwrap(), expected);
    }

    #[test]
    fn formats_csv_output() {
        let result = QueryResult {
            columns: vec!["name".to_string(), "line".to_string()],
            rows: vec![vec![Value::String("main".to_string()), Value::from(4)]],
        };

        let mut output = Vec::new();
        format_response(&mut output, "csv", &result).expect("format should succeed");

        let expected = "name,line\nmain,4\n";
        assert_eq!(String::from_utf8(output).unwrap(), expected);
    }

    #[test]
    fn handles_empty_columns() {
        let result = QueryResult {
            columns: vec![],
            rows: vec![],
        };

        let mut output = Vec::new();
        format_response(&mut output, "table", &result).expect("format should succeed");
        assert!(String::from_utf8(output).unwrap().is_empty());
    }
}
